### DISCUSSION TITLE
Unable to Inject IMemoryStore in my project using VolatileMemoryStore.

### DISCUSSION THREAD

**Discussion by Elio-Khater** [microsoft/semantic-kernel#8625] [Q&A] [answered]
using the example provided in the class PluginSelectionWithFilters.cs, I tried to apply the same logic to my app.
but when I try to inject IMemoryStore as done the in documentation : 

builder.Services.AddSingleton<IMemoryStore, VolatileMemoryStore>();

I can't find VolatileMemoryStore. I tried updating all my packages, but it didn't work.

Anyone has any ideas?

**Comment by evchaki** [comment]
@westey-m  - can you double check this is still available?

**Comment by westey-m** [comment]
@Elio-Khater, do you have a code snippet that you can share that reproduces the issue?  E.g. can you show how/where are you consuming IMemoryStore and how you are resolving the consuming classes, etc?

**Reply by Elio-Khater** [reply]
@westey-m I am using this IMemoryStore in pluginStore and functionProvider the same way it is used in PluginSelectionWithFilters.cs

then im trying to inject it here : 

 public static IServiceCollection AddKernel(this IServiceCollection services,string baseUrl, string apiKey, IList<LLmDeploymentConfig> llmDeploymentConfigs, string provider)
 {
     services.AddScoped((providers) =>
     {
         var builder = Kernel.CreateBuilder();
        
         builder.Services.AddSingleton(providers.GetRequiredService<IMediator>());
         builder.Services.AddSingleton<IPluginStore, PluginStore>();
         builder.Services.AddSingleton<IFunctionKeyProvider, FunctionKeyProvider>();
         builder.Services.AddSingleton<IFunctionProvider, FunctionProvider>();
         builder.Services.AddSingleton<IMemoryStore, VolatileMemoryStore>();
        ....
       }
 }
        
        with VolatileMemoryStore not found.

**Reply by westey-m** [reply]
Thanks for the extra information @Elio-Khater.  Can you also show how you are constructing the function provider?

**Comment by Elio-Khater** [comment]
@westey-m, here is my FunctionProvider.cs : 

public class FunctionProvider : IFunctionProvider
{
    private readonly ITextEmbeddingGenerationService _textEmbeddingGenerationService;
    private readonly IMemoryStore _memoryStore;
    private readonly IFunctionKeyProvider _functionKeyProvider;

    public FunctionProvider(
        ITextEmbeddingGenerationService textEmbeddingGenerationService,
        IMemoryStore memoryStore,
        IFunctionKeyProvider functionKeyProvider)
    {
        _textEmbeddingGenerationService = textEmbeddingGenerationService;
        _memoryStore = memoryStore;
        _functionKeyProvider = functionKeyProvider;
    }

    public async Task<List<KernelFunction>> GetBestFunctionsAsync(
        string collectionName,
        string request,
        KernelPluginCollection plugins,
        int numberOfBestFunctions)
    {
        var requestEmbedding = await _textEmbeddingGenerationService.GenerateEmbeddingAsync(request);

        var memoryRecordKeys = await _memoryStore
            .GetNearestMatchesAsync(collectionName, requestEmbedding, limit: numberOfBestFunctions)
            .Select(l => l.Item1.Key)
            .ToListAsync();

        return plugins
            .SelectMany(plugin => plugin)
            .Where(function => memoryRecordKeys.Contains(_functionKeyProvider.GetFunctionKey(function)))
            .ToList();
    }
}

**Reply by westey-m** [reply, marked as answer]
Thanks @Elio-Khater.  The issue is going to be somewhere around how the function provider is constructed/resolved or how whatever consumes it is constructed/resolved.  I don't see anything wrong with how you are registering the Memory Store or Function provider, so the issue must be around how it's constructed/resolved from DI.  If you can share more information about how you are doing that, we might be able to spot the issue.
